### PR TITLE
Jf withdraw

### PR DESCRIPTION
**Pull request by mike1729**
Native

- ark/withdraw/keygen     time:   [1.4768 s 1.4776 s 1.4785 s]
- ark/withdraw/prover     time:   [1.2182 s 1.2185 s 1.2187 s]
- ark/withdraw/verifier   time:   [4.5749 ms 4.5755 ms 4.5761 ms]
- jf/withdraw/keygen      time:   [716.69 ms 717.28 ms 717.85 ms]
- jf/withdraw/prover      time:   [905.92 ms 907.23 ms 908.49 ms]
- jf/withdraw/verifier    time:   [17.795 ms 17.799 ms 17.803 ms]

Wasm

  -  "jf_withdraw" key generation.        37.38s
  -  "ark_withdraw" key generation.     9.06s
  -  "jf_withdraw" prover.                     37.445s
  -  "ark_withdraw" prover.                  8.074s

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -118,3 +132,89 @@ impl Relation {
...
+    ) {
+        let input = relation().public_input();
+        assert!(
+            PlonkKzgSnark::<Curve>::verify::<StandardTranscript>(&vk, &input, &proof, None,)
+                .is_ok()
+        )
+    }
+}
+
+fn relation() -> WithdrawRelation {
```
Comment (deuszx) [inline src/lib.rs:167]: That's jellyfish, right?

**Review by deuszx — APPROVED** [review]
I'd just split `src/lib.rs` content into `ark.rs` and `jellyfish.rs` files.

File: src/lib.rs [hunk 2]
```diff
@@ -118,3 +132,89 @@ impl Relation {
...
 }
+
+pub enum JfRelation {
+    Withdraw,
+}
+
+impl JfRelation {
+    pub fn generate_keys(&self) -> (JfProvingKey<Curve>, JfVerifyingKey<Curve>) {
+        let rng = &mut jf_utils::test_rng();
+        let srs = generate_srs(17_000, rng).unwrap();
```
Comment (pmikolajczyk41) [inline src/lib.rs:143]: this shouldn't be benchmarked

**Review by pmikolajczyk41 — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 2 above)
Comment (mike1729) [reply src/lib.rs]: Rand generators are usually initialized once and getting a handle is almost free, it takes nanoseconds. Good point with srs. I've noticed also that for many operations, we generate circuit from scratch, e.g. for proof generation.

File: src/lib.rs (hunk 1 above)
Comment (mike1729) [reply src/lib.rs]: Right.

File: src/lib.rs (hunk 2 above)
Comment (mike1729) [reply src/lib.rs]: Fixed.

File: tests/key_generation.rs [hunk 3]
```diff
@@ -27,11 +27,45 @@ fn bench(relation_name: &str) {
...
+            "{:?} key generation. Relation performance: {:?}",
+            operation,
+            (elapsed / REPEAT as u32)
+        )
+        .into(),
+    );
+}
+
+#[wasm_bindgen_test]
+fn jf_bench_withdraw() {
```
Comment (pmikolajczyk41) [inline tests/key_generation.rs:58]: don't we miss proving?

Comment (mike1729) [reply tests/key_generation.rs:58]: It's in separate file tests/proving.rs

**Review by deuszx — APPROVED** [review]

File: tests/key_generation.rs [hunk 4]
```diff
@@ -27,11 +27,45 @@ fn bench(relation_name: &str) {
...
+
+    let job = || relation.generate_circuit();
+    jf_bench("circuit generation", job);
+
+    let job = || {let _ =relation.generate_srs();};
+    jf_bench("srs generation", job);
+
+    let srs = relation.generate_srs();
+
+    let job = ||{let _ =relation.generate_keys(&srs);};
```
[the remark this answers is not in the document]
Comment (mike1729) [reply tests/key_generation.rs]: Fixed.

**Pull request closed without merging**